src/box/schema.cc

```diff
@@ -487,3 +489,307 @@ schema_find_name(enum schema_object_type type, uint32_t object_id)
 	return NULL;
 }
 
+/**
+ * Drop the constraint or foreign key. If fieldno is equal to UINT32_MAX then
+ * the tuple constraint or tuple foreign key is dropped, otherwise the field
+ * constraint or field foreign key is dropped.
+ */
+int
+constraint_drop(uint32_t space_id, const char *name, uint32_t name_len,
+		uint32_t fieldno, enum tuple_constraint_type type)
+{
+	char key[16];
+	char *key_end = mp_encode_array(key, 1);
+	key_end = mp_encode_uint(key_end, space_id);
+	struct tuple *tuple;
+	if (box_index_get(BOX_SPACE_ID, 0, key, key_end, &tuple) != 0)
+		return -1;
+	assert(tuple != NULL);
+	const char *data = tuple_data(tuple);
+	const char *data_end = data;
+	mp_next(&data_end);
+
+	const char *str;
+	if (fieldno != UINT32_MAX) {
+		str = tuple_field(tuple, BOX_SPACE_FIELD_FORMAT);
+		assert(str != NULL);
+		mp_decode_array(&str);
+		for (uint32_t i = 0; i < fieldno; ++i)
+			mp_next(&str);
+	} else {
+		str = tuple_field(tuple, BOX_SPACE_FIELD_OPTS);
+		assert(str != NULL);
+	}
+
+	const char *prev = str;
+	uint32_t count = mp_decode_map(&str);
+	const char *mid = str;
+	uint32_t id = 0;
+	const char *constraints = str;
+	const char *constraint_type =
+		type == CONSTR_FKEY ? "foreign_key" : "constraint";
+	for (; id < count; ++id) {
+		uint32_t len;
+		const char *tmp = mp_decode_str(&str, &len);
+		if (len == strlen(constraint_type) &&
+		    memcmp(tmp, constraint_type, len) == 0)
+			break;
+		mp_next(&str);
+		constraints = str;
+	}
+	if (id == count) {
+		int errcode = type == CONSTR_FKEY ? ER_NO_SUCH_FOREIGN_KEY :
+			      ER_NO_SUCH_CONSTRAINT;
+		diag_set(ClientError, errcode, name,
+			 space_name(space_by_id(space_id)));
+		return -1;
+	}
+
+	uint32_t constraint_count = mp_decode_map(&str);
+	const char *constraint_begin = str;
+	const char *constraint = str;
+	uint32_t constraint_id = 0;
+	for (; constraint_id < constraint_count; ++constraint_id) {
+		uint32_t len;
+		const char *tmp = mp_decode_str(&str, &len);
+		if (name_len == len && memcmp(tmp, name, len) == 0)
+			break;
+		mp_next(&str);
+		constraint = str;
+	}
+	if (constraint_id == constraint_count) {
+		int errcode = type == CONSTR_FKEY ? ER_NO_SUCH_FOREIGN_KEY :
+			      ER_NO_SUCH_CONSTRAINT;
+		diag_set(ClientError, errcode, name,
+			 space_name(space_by_id(space_id)));
+		return -1;
+	}
+	const char *post = constraint;
```

Comment:
Instead of all those tricky decodings could try to re-use `space_def_new_from_tuple()`. Although you would need to implement `space_def_to_tuple()` or `space_def_to_msgpack()`. Then you would decode a `space_def`, do the needed changes without the msgpack tricks, encode it back, and use `box_replace()` like now. Up to you if you want to try.

Reply:
I used your suggestion about `box_update()`.